<?xml version='1.0'?>
<sdf version='1.9'>
  <world name="Line_track">
  
    <physics name="1ms" type="ignored">
      <max_step_size>0.001</max_step_size>
      <real_time_factor>1.0</real_time_factor>
    </physics>
    <plugin
      filename="ignition-gazebo-physics-system"
      name="gz::sim::systems::Physics">
    </plugin>
    <plugin
      filename="ignition-gazebo-user-commands-system"
      name="gz::sim::systems::UserCommands">
    </plugin>
    <plugin
      filename="ignition-gazebo-scene-broadcaster-system"
      name="gz::sim::systems::SceneBroadcaster">
    </plugin>
    <plugin name='gz::sim::systems::Sensors' filename='ignition-gazebo-sensors-system'>
      <render_engine>ogre2</render_engine>
    </plugin>
<!--
    <gui fullscreen='false'>
      <plugin name='3D View' filename='MinimalScene'>
        <ignition-gui>
          <title>3D View</title>
          <property type='bool' key='showTitleBar'>false</property>
          <property type='string' key='state'>docked</property>
        </ignition-gui>
        <engine>ogre2</engine>
        <scene>scene</scene>
        <ambient_light>0.4 0.4 0.4</ambient_light>
        <background_color>0.8 0.8 0.8</background_color>
        <camera_pose>-6 0 6 0 0.5 0</camera_pose>
      </plugin>
      <plugin name='Entity context menu' filename='EntityContextMenuPlugin'>
        <ignition-gui>
          <property key='state' type='string'>floating</property>
          <property key='width' type='double'>5</property>
          <property key='height' type='double'>5</property>
          <property key='showTitleBar' type='bool'>false</property>
        </ignition-gui>
      </plugin>
      <plugin name='Scene Manager' filename='GzSceneManager'>
        <ignition-gui>
          <property key='resizable' type='bool'>false</property>
          <property key='width' type='double'>5</property>
          <property key='height' type='double'>5</property>
          <property key='state' type='string'>floating</property>
          <property key='showTitleBar' type='bool'>false</property>
        </ignition-gui>
      </plugin>
      <plugin name='Interactive view control' filename='InteractiveViewControl'>
        <ignition-gui>
          <property key='resizable' type='bool'>false</property>
          <property key='width' type='double'>5</property>
          <property key='height' type='double'>5</property>
          <property key='state' type='string'>floating</property>
          <property key='showTitleBar' type='bool'>false</property>
        </ignition-gui>
      </plugin>
      <plugin name='Camera Tracking' filename='CameraTracking'>
        <ignition-gui>
          <property key='resizable' type='bool'>false</property>
          <property key='width' type='double'>5</property>
          <property key='height' type='double'>5</property>
          <property key='state' type='string'>floating</property>
          <property key='showTitleBar' type='bool'>false</property>
        </ignition-gui>
      </plugin>
      <plugin name='World control' filename='WorldControl'>
        <ignition-gui>
          <title>World control</title>
          <property type='bool' key='showTitleBar'>false</property>
          <property type='bool' key='resizable'>false</property>
          <property type='double' key='height'>72</property>
          <property type='double' key='width'>121</property>
          <property type='double' key='z'>1</property>
          <property type='string' key='state'>floating</property>
          <anchors target='3D View'>
            <line own='left' target='left'/>
            <line own='bottom' target='bottom'/>
          </anchors>
        </ignition-gui>
        <play_pause>true</play_pause>
        <step>true</step>
        <start_paused>true</start_paused>
        <use_event>true</use_event>
      </plugin>
      <plugin name='World stats' filename='WorldStats'>
        <ignition-gui>
          <title>World stats</title>
          <property type='bool' key='showTitleBar'>false</property>
          <property type='bool' key='resizable'>false</property>
          <property type='double' key='height'>110</property>
          <property type='double' key='width'>290</property>
          <property type='double' key='z'>1</property>
          <property type='string' key='state'>floating</property>
          <anchors target='3D View'>
            <line own='right' target='right'/>
            <line own='bottom' target='bottom'/>
          </anchors>
        </ignition-gui>
        <sim_time>true</sim_time>
        <real_time>true</real_time>
        <real_time_factor>true</real_time_factor>
        <iterations>true</iterations>
      </plugin>
      <plugin name='Component inspector' filename='ComponentInspector'>
        <ignition-gui>
          <property type='string' key='state'>docked</property>
        </ignition-gui>
      </plugin>
      <plugin name='Entity tree' filename='EntityTree'>
        <ignition-gui>
          <property type='string' key='state'>docked</property>
        </ignition-gui>
      </plugin>  
      
      <plugin filename="WorldControl" name="World control">
        <ignition-gui>
          <title>World control</title>
          <property type="bool" key="showTitleBar">false</property>
          <property type="bool" key="resizable">false</property>
          <property type="double" key="height">72</property>
          <property type="double" key="width">121</property>
          <property type="double" key="z">1</property>

          <property type="string" key="state">floating</property>
          <anchors target="3D View">
            <line own="left" target="left"/>
            <line own="bottom" target="bottom"/>
          </anchors>
        </ignition-gui>

        <play_pause>true</play_pause>
        <step>true</step>
        <start_paused>true</start_paused>

      </plugin>
      
      <plugin filename="Teleop" name="Teleop">
        <ignition-gui>
          <property type="string" key="state">docked</property>
        </ignition-gui>
        <topic>/cmd_vel</topic>
      </plugin>
      
    </gui> 
-->
    <light type="directional" name="sun">
      <cast_shadows>true</cast_shadows>
      <pose>0 0 10 0 0 0</pose>
      <diffuse>1 1 1 1</diffuse>
      <specular>0.5 0.5 0.5 1</specular>
      <attenuation>
        <range>1000</range>
        <constant>0.9</constant>
        <linear>0.01</linear>
        <quadratic>0.001</quadratic>
      </attenuation>
      <direction>-0.5 0.1 -0.9</direction>
    </light>

    <model name="ground_plane">
      <static>true</static>
      <link name="link">
        <collision name="collision">
          <geometry>
            <plane>
              <normal>0 0 1</normal>
              <size>1 1</size>
            </plane>
          </geometry>
          <surface>
            <friction>
              <ode>
                <mu>50</mu>
              </ode>
            </friction>
          </surface>
        </collision>
        <visual name="visual">
          <geometry>
            <plane>
              <normal>0 0 1</normal>
              <size>6 6</size>
            </plane>
          </geometry>
          <material>
            <ambient>0.1 0.1 0.1 0.1</ambient>
            <diffuse>0.7 0.7 0.7 1</diffuse>
   	    <pbr>
   	      <metal>
   	        <albedo_map>track_line_v1.png</albedo_map>
   	        <normal_map>track_line_v1.png</normal_map>
   	      </metal>
   	    </pbr>
          </material>
        </visual>
      </link>
      <static>1</static>
    </model>
    
    <model name="JetTec_Robot">
      <pose>-1.0 -0.90 0.15 0 0 0</pose>

      <!-- Châssis principal -->
      <link name="chassis">
        <pose>1 -1.0 0.06 0 0 0</pose>
        <inertial>
          <mass>2.5</mass>
          <inertia>
            <ixx>0.126164</ixx>
            <ixy>0</ixy>
            <ixz>0</ixz>
            <iyy>0.416519</iyy>
            <iyz>0</iyz>
            <izz>0.481014</izz>
          </inertia>
        </inertial>
        <visual name="visual">
          <geometry>
            <box>
              <size>0.53 0.15 0.12</size>
            </box>
          </geometry>
          <material>
            <ambient>0.5 0.5 1.0 1</ambient>
            <diffuse>0.5 0.5 1.0 1</diffuse>
            <specular>0.0 0.0 1.0 1</specular>
          </material>
        </visual>
        <collision name="collision">
          <geometry>
            <box>
              <size>0.53 0.15 0.12</size>
            </box>
          </geometry>
        </collision>
      </link>
      
      <!-- Lien pour la caméra , avec corps physique et visuel -->
      
      <link name='rgbd'>
        <pose>1.25 -1.0 0.20 0 -0 0</pose>
        <inertial>
          <mass>0.10000000000000001</mass>
          <inertia>
            <ixx>0.00016666700000000001</ixx>
            <iyy>0.00016666700000000001</iyy>
            <izz>0.00016666700000000001</izz>
            <ixy>0</ixy>
            <ixz>0</ixz>
            <iyz>0</iyz>
          </inertia>
          <pose>0 0 0 0 -0 0</pose>
        </inertial>
        <collision name='collision'>
          <geometry>
            <box>
              <size>0.1 0.1 0.1</size>
            </box>
          </geometry>
          <surface>
            <friction>
              <ode/>
            </friction>
            <bounce/>
            <contact/>
          </surface>
        </collision>
        <visual name='visual'>
          <geometry>
            <box>
              <size>0.1 0.1 0.1</size>
            </box>
          </geometry>
        </visual>
        <sensor name='rgbd_camera' type='rgbd_camera'>
          <camera>
            <horizontal_fov>2</horizontal_fov>
            <image>
              <width>800</width>
              <height>600</height>
            </image>
            <clip>
              <near>0.10000000000000001</near>
              <far>100</far>
            </clip>
          </camera>
          <always_on>true</always_on>
          <update_rate>10</update_rate>
          <visualize>true</visualize>
          <topic>/rgbd_camera</topic>
          <enable_metrics>true</enable_metrics>
        </sensor>
        <enable_wind>false</enable_wind>
      </link>

      <!-- Roues avant gauche -->
      <link name="front_left_wheel">
        <pose relative_to="chassis">0.265 0.14 -0.06 -1.5707 0 0</pose>
        <inertial>
          <mass>0.3</mass>
          <inertia>
            <ixx>0.145833</ixx>
            <ixy>0</ixy>
            <ixz>0</ixz>
            <iyy>0.145833</iyy>
            <iyz>0</iyz>
            <izz>0.125</izz>
          </inertia>
        </inertial>
        <visual name="visual">
          <geometry>
            <cylinder>
              <length>0.06</length>
              <radius>0.075</radius>
            </cylinder>
          </geometry>
          <material>
            <ambient>0.2 0.2 0.2 1</ambient>
            <diffuse>0.2 0.2 0.2 1</diffuse>
            <specular>0.2 0.2 0.2 1</specular>
          </material>
        </visual>
        <collision name="collision">
          <geometry>
            <cylinder>
              <length>0.06</length>
              <radius>0.075</radius>
            </cylinder>
          </geometry>
          <surface>
            <friction>
              <ode>
                <mu>0.5</mu>
                <mu2>1.0</mu2>
                <fdir1>0 0 1</fdir1>
              </ode>
            </friction>
          </surface>
        </collision>
      </link>

      <!-- Roues arrière gauche -->
      <link name="rear_left_wheel">
        <pose relative_to="chassis">-0.265 0.14 -0.06 -1.5707 0 0</pose>
        <inertial>
          <mass>0.3</mass>
          <inertia>
            <ixx>0.145833</ixx>
            <ixy>0</ixy>
            <ixz>0</ixz>
            <iyy>0.145833</iyy>
            <iyz>0</iyz>
            <izz>0.125</izz>
          </inertia>
        </inertial>
        <visual name="visual">
          <geometry>
            <cylinder>
              <length>0.06</length>
              <radius>0.075</radius>
            </cylinder>
          </geometry>
          <material>
            <ambient>0.2 0.2 0.2 1</ambient>
            <diffuse>0.2 0.2 0.2 1</diffuse>
            <specular>0.2 0.2 0.2 1</specular>
          </material>
        </visual>
        <collision name="collision">
          <geometry>
            <cylinder>
              <length>0.06</length>
              <radius>0.075</radius>
            </cylinder>
          </geometry>
          <surface>
            <friction>
              <ode>
                <mu>0.5</mu>
                <mu2>1.0</mu2>
                <fdir1>0 0 1</fdir1>
              </ode>
            </friction>
          </surface>
        </collision>
      </link>

      <!-- Roues avant droite -->
      <link name="front_right_wheel">
        <pose relative_to="chassis">0.265 -0.14 -0.06 -1.5707 0 0</pose>
        <inertial>
          <mass>0.3</mass>
          <inertia>
            <ixx>0.145833</ixx>
            <ixy>0</ixy>
            <ixz>0</ixz>
            <iyy>0.145833</iyy>
            <iyz>0</iyz>
            <izz>0.125</izz>
          </inertia>
        </inertial>
        <visual name="visual">
          <geometry>
            <cylinder>
              <length>0.06</length>
              <radius>0.075</radius>
            </cylinder>
          </geometry>
          <material>
            <ambient>0.2 0.2 0.2 1</ambient>
            <diffuse>0.2 0.2 0.2 1</diffuse>
            <specular>0.2 0.2 0.2 1</specular>
          </material>
        </visual>
        <collision name="collision">
          <geometry>
            <cylinder>
              <length>0.06</length>
              <radius>0.075</radius>
            </cylinder>
          </geometry>
          <surface>
            <friction>
              <ode>
                <mu>0.5</mu>
                <mu2>1.0</mu2>
                <fdir1>0 0 1</fdir1>
              </ode>
            </friction>
          </surface>
        </collision>
      </link>

      <!-- Roues arrière droite -->
      <link name="rear_right_wheel">
        <pose relative_to="chassis">-0.265 -0.14 -0.06 -1.5707 0 0</pose>
        <inertial>
          <mass>0.3</mass>
          <inertia>
            <ixx>0.145833</ixx>
            <ixy>0</ixy>
            <ixz>0</ixz>
            <iyy>0.145833</iyy>
            <iyz>0</iyz>
            <izz>0.125</izz>
          </inertia>
        </inertial>
        <visual name="visual">
          <geometry>
            <cylinder>
              <length>0.06</length>
              <radius>0.075</radius>
            </cylinder>
          </geometry>
          <material>
            <ambient>0.2 0.2 0.2 1</ambient>
            <diffuse>0.2 0.2 0.2 1</diffuse>
            <specular>0.2 0.2 0.2 1</specular>
          </material>
        </visual>
        <collision name="collision">
          <geometry>
            <cylinder>
              <length>0.06</length>
              <radius>0.075</radius>
            </cylinder>
          </geometry>
          <surface>
            <friction>
              <ode>
                <mu>0.5</mu>
                <mu2>1.0</mu2>
                <fdir1>0 0 1</fdir1>
              </ode>
            </friction>
          </surface>
        </collision>
      </link>

      <!-- Liens de direction pour les roues avant -->
      <link name="front_left_wheel_steering_link">
        <pose relative_to="chassis">0.265 0.07 -0.02 0 0 0</pose>
        <inertial>
          <mass>0.5</mass>
          <inertia>
            <ixx>0.0153</ixx>
            <iyy>0.025</iyy>
            <izz>0.0153</izz>
          </inertia>
        </inertial>
        <visual name="steering_link_visual">
          <pose>0 0 0 0 0 0</pose>
          <geometry>
            <cylinder>
              <length>0.05</length>
              <radius>0.015</radius>
            </cylinder>
          </geometry>
          <material>
            <ambient>1 1 1</ambient>
            <diffuse>1 1 1</diffuse>
          </material>
        </visual>
      </link>

      <link name="front_right_wheel_steering_link">
        <pose relative_to="chassis">0.265 -0.07 -0.02 0 0 0</pose>
        <inertial>
          <mass>0.5</mass>
          <inertia>
            <ixx>0.0153</ixx>
            <iyy>0.025</iyy>
            <izz>0.0153</izz>
          </inertia>
        </inertial>
        <visual name="steering_link_visual">
          <pose>0 0 0 0 0 0</pose>
          <geometry>
            <cylinder>
              <length>0.05</length>
              <radius>0.015</radius>
            </cylinder>
          </geometry>
          <material>
            <ambient>1 1 1</ambient>
            <diffuse>1 1 1</diffuse>
          </material>
        </visual>
      </link>

      <!-- Joints pour la direction -->
      <joint name="front_left_wheel_steering_joint" type="revolute">
        <child>front_left_wheel_steering_link</child>
        <parent>chassis</parent>
        <axis>
          <xyz>0 0 1</xyz>
          <limit>
            <lower>-0.6</lower>
            <upper>0.6</upper>
            <velocity>1.0</velocity>
            <effort>25</effort>
          </limit>
          <use_parent_model_frame>1</use_parent_model_frame>
        </axis>
      </joint>

      <joint name="front_right_wheel_steering_joint" type="revolute">
        <parent>chassis</parent>
        <child>front_right_wheel_steering_link</child>
        <axis>
          <xyz>0 0 1</xyz>
          <limit>
            <lower>-0.6</lower>
            <upper>0.6</upper>
            <velocity>1.0</velocity>
            <effort>25</effort>
          </limit>
        </axis>
      </joint>

      <!-- Joints pour les roues -->
      <joint name="front_left_wheel_joint" type="revolute">
        <parent>front_left_wheel_steering_link</parent>
        <child>front_left_wheel</child>
        <axis>
          <xyz>0 0 1</xyz>
          <limit>
            <lower>-1.79769e+308</lower>
            <upper>1.79769e+308</upper>
          </limit>
        </axis>
      </joint>

      <joint name="front_right_wheel_joint" type="revolute">
        <parent>front_right_wheel_steering_link</parent>
        <child>front_right_wheel</child>
        <axis>
          <xyz>0 0 1</xyz>
          <limit>
            <lower>-1.79769e+308</lower>
            <upper>1.79769e+308</upper>
          </limit>
        </axis>
      </joint>

      <joint name="rear_left_wheel_joint" type="revolute">
        <parent>chassis</parent>
        <child>rear_left_wheel</child>
        <axis>
          <xyz>0 0 1</xyz>
          <limit>
            <lower>-1.79769e+308</lower>
            <upper>1.79769e+308</upper>
          </limit>
        </axis>
      </joint>

      <joint name="rear_right_wheel_joint" type="revolute">
        <parent>chassis</parent>
        <child>rear_right_wheel</child>
        <axis>
          <xyz>0 0 1</xyz>
          <limit>
            <lower>-1.79769e+308</lower>
            <upper>1.79769e+308</upper>
          </limit>
        </axis>
      </joint>
      
      <!-- Joint pour fixer la caméra au châssis -->
      <joint name="camera_joint" type="fixed">
        <parent>chassis</parent>
        <child>rgbd</child>
        <pose>0 0 0 0 0 0</pose>
      </joint>

      <!-- Plugin de commande Ackermann -->
      <plugin filename="ignition-gazebo-ackermann-steering-system" name="gz::sim::systems::AckermannSteering">
 	<topic>/cmd_vel</topic>
  	<left_joint>front_left_wheel_joint</left_joint>
  	<left_joint>rear_left_wheel_joint</left_joint>
  	<right_joint>front_right_wheel_joint</right_joint>
  	<right_joint>rear_right_wheel_joint</right_joint>
  	<left_steering_joint>front_left_wheel_steering_joint</left_steering_joint>
  	<right_steering_joint>front_right_wheel_steering_joint</right_steering_joint>
  	<kingpin_width>1.0</kingpin_width>
 	<steering_limit>0.5</steering_limit>
  	<wheel_base>1.0</wheel_base>
  	<wheel_separation>1.25</wheel_separation>
  	<wheel_radius>0.3</wheel_radius>
  	<min_velocity>-3</min_velocity>
  	<max_velocity>3</max_velocity>
  	<min_acceleration>-3</min_acceleration>
  	<max_acceleration>3</max_acceleration>
      </plugin>

      
      <!-- joint state publisher plugin -->
        <plugin
            filename="gz-sim-joint-state-publisher-system"
            name="gz::sim::systems::JointStatePublisher">
            <joint_name>front_left_wheel_steering_joint</joint_name>
            <joint_name>front_right_wheel_steering_joint</joint_name>
            <joint_name>front_left_wheel_joint</joint_name>
            <joint_name>front_right_wheel_joint</joint_name>
            <joint_name>rear_left_wheel_joint</joint_name>
            <joint_name>rear_right_wheel_joint</joint_name>
            <joint_name>camera_joint</joint_name>
        </plugin>
      
    </model>
    
  </world>
</sdf>
